components/cop_codegen/src/rpn_function.rs

```diff
@@ -62,3 +71,579 @@ impl RpnFunctionOpts {
+            }
+        }
+    }
+
+    fn generate_real_fn_trait_impl(&self) -> TokenStream {
+        let mut generics = self.item_fn.decl.generics.clone();
+        generics
+            .params
+            .push(parse_str::<GenericParam>("'arg_").unwrap());
+        let mut tp = NULL.to_string();
+        for (arg_index, arg_type) in self.arg_types.iter().enumerate().rev() {
+            let arg_name = format!("Arg{}_", arg_index);
+            let generic_param = format!(
+                "{}: {}<Type = &'arg_ Option<{}>>",
+                arg_name,
+                RPN_FN_ARG,
+                arg_type.into_token_stream()
+            );
+            generics.params.push(parse_str(&generic_param).unwrap());
+            tp = format!("{}<{}, {}>", ARG, arg_name, tp);
+        }
+        let fn_ident = &self.item_fn.ident;
+        let fn_trait_ident = &self.fn_trait_ident;
+        let tp = parse_str::<Type>(&tp).unwrap();
+        let (_, ty_generics, _) = self.item_fn.decl.generics.split_for_impl();
+        let (impl_generics, _, where_clause) = generics.split_for_impl();
+        let meta = &self.meta;
+        let extract = (0..self.arg_types.len())
+            .map(|i| syn::parse_str::<Ident>(&format!("arg{}", i)).unwrap());
+        let call_arg = extract.clone();
+        let ty_generics_turbofish = ty_generics.as_turbofish();
+        let (ctx_type, payload_type, result_type) = (ctx_type(), payload_type(), result_type());
+        quote! {
+            impl #impl_generics #fn_trait_ident #ty_generics for #tp #where_clause {
+                default fn eval(
+                    self,
+                    rows: usize,
+                    ctx: &mut #ctx_type,
+                    payload: #payload_type,
+                ) -> #result_type {
+                    let arg = &self;
+                    let mut result = Vec::with_capacity(rows);
+                    for row in 0..rows {
+                        #(let (#extract, arg) = arg.extract(row));*;
+                        result.push( #fn_ident #ty_generics_turbofish ( #(#meta),* #(#call_arg),* )?);
+                    }
+                    Ok(crate::coprocessor::codec::data_type::Evaluable::into_vector_value(result))
+                }
+            }
+        }
+    }
+
+    fn generate_evaluator(&self) -> TokenStream {
+        let (impl_generics, ty_generics, where_clause) =
+            self.item_fn.decl.generics.split_for_impl();
+        let evaluator_ident = &self.evaluator_ident;
+        let fn_trait_ident = &self.fn_trait_ident;
+        let ty_generics_turbofish = ty_generics.as_turbofish();
+        let evaluator_def = if self.item_fn.decl.generics.params.empty_or_trailing() {
+            quote! { pub struct #evaluator_ident; }
+        } else {
+            let types = format!("{}", ty_generics.clone().into_token_stream());
+            let phantom_generic: TokenStream =
+                parse_str(&format!("<({})>", &types[1..types.len() - 1])).unwrap();
```

Comment:
```rust
let types_middle = &types[1..(types.len() - 1)];
let phantom_generic = quote!{ <(#(#types_middle),*)> };
```

Reply:
`types` here is a string, we cannot write this way...I decide to use `generics.type_params().map(|type_param| type_param.ident)`, and put it in the `quote!`. Lifetime won't appear here.